### PR TITLE
Sync with tier4 master

### PR DESCRIPTION
**Pull request by wep21**

### CODE REVIEW

**Comment by kenji-miyake** [comment]
@wep21 How about splitting PR (or commit) so that we can distinguish what is for updating ROS APIs and what is for our customization?

**Comment by wep21** [comment]
@kenji-miyake I split PR into https://github.com/ros-perception/pointcloud_to_laserscan/pull/56 and this.

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,14 +5,21 @@ find_package(ament_cmake REQUIRED)
...
 find_package(laser_geometry REQUIRED)
 find_package(message_filters REQUIRED)
+find_package(pcl_conversions REQUIRED)
 find_package(rclcpp REQUIRED)
 find_package(rclcpp_components REQUIRED)
 find_package(sensor_msgs REQUIRED)
 find_package(tf2 REQUIRED)
 find_package(tf2_ros REQUIRED)
 find_package(tf2_sensor_msgs REQUIRED)
+find_package(visualization_msgs REQUIRED)
```
Comment (kenji-miyake) [inline CMakeLists.txt]: Will we use `ament_cmake_auto` and send a PR to upstream?

File: src/pointcloud_to_laserscan_node.cpp [hunk 2]
```diff
@@ -225,9 +243,80 @@ void PointCloudToLaserScanNode::cloudCallback(
...
     }
   }
-  pub_->publish(std::move(scan_msg));
+
+  laserscan_pub_->publish(std::move(scan_msg));
+
+  pcl::PointCloud<pcl::PointXYZ> pcl_pointcloud;
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(*cloud_msg, "x"), iter_y(*cloud_msg, "y"),
+    iter_z(*cloud_msg, "z");
+  for (size_t i = 0; i < v_pointcloud_index.size(); ++i) {
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: Can we replace this with `range-based-for`?

File: src/pointcloud_to_laserscan_node.cpp [hunk 3]
```diff
@@ -225,9 +243,80 @@ void PointCloudToLaserScanNode::cloudCallback(
...
       scan_msg->ranges[index] = range;
+      v_pointcloud_index.at(index) = pointcloud_index;
     }
   }
-  pub_->publish(std::move(scan_msg));
+
+  laserscan_pub_->publish(std::move(scan_msg));
+
+  pcl::PointCloud<pcl::PointXYZ> pcl_pointcloud;
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(*cloud_msg, "x"), iter_y(*cloud_msg, "y"),
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: Declaring multiple variables in the same line isn't allowed.
Please see A7-1-7 in https://www.autosar.org/fileadmin/user_upload/standards/adaptive/17-03/AUTOSAR_RS_CPP14Guidelines.pdf.

File: src/pointcloud_to_laserscan_node.cpp [hunk 4]
```diff
@@ -225,9 +243,80 @@ void PointCloudToLaserScanNode::cloudCallback(
...
+    if (v_pointcloud_index.at(i) != no_data) {
+      pcl::PointXYZ point;
+      point.x = *(iter_x + v_pointcloud_index.at(i));
+      point.y = *(iter_y + v_pointcloud_index.at(i));
+      point.z = *(iter_z + v_pointcloud_index.at(i));
+      pcl_pointcloud.push_back(point);
+    }
+  }
+  pcl::toROSMsg(pcl_pointcloud, pointcloud_output);
+  auto pointcloud_output_ptr
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: [nits] const?

File: src/pointcloud_to_laserscan_node.cpp [hunk 5]
```diff
@@ -152,6 +156,17 @@ void PointCloudToLaserScanNode::cloudCallback(
...
+  // build pointcloud2 output
+  sensor_msgs::msg::PointCloud2 pointcloud_output;
+  const int no_data = -1;
+  std::vector<int> v_pointcloud_index;
+
+  // build ray output
+  visualization_msgs::msg::Marker ray_output;
+
+  // build stixel output
+  visualization_msgs::msg::MarkerArray stixel_output;
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: I think it's better to move these variables(pointcloud_output, ray_output, stixel_output) just before the part they are used in order to reduce the scope.

File: src/pointcloud_to_laserscan_node.cpp [hunk 6]
```diff
@@ -225,9 +243,80 @@ void PointCloudToLaserScanNode::cloudCallback(
...
+      pcl::PointXYZ point;
+      point.x = *(iter_x + v_pointcloud_index.at(i));
+      point.y = *(iter_y + v_pointcloud_index.at(i));
+      point.z = *(iter_z + v_pointcloud_index.at(i));
+      pcl_pointcloud.push_back(point);
+    }
+  }
+  pcl::toROSMsg(pcl_pointcloud, pointcloud_output);
+  auto pointcloud_output_ptr
+    = std::make_unique<sensor_msgs::msg::PointCloud2>(pointcloud_output);
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: [nits] `sensor_msgs::msg::PointCloud2` can be abbreviated? Then you can write in one line.

File: src/pointcloud_to_laserscan_node.cpp [hunk 7]
```diff
@@ -225,9 +243,80 @@ void PointCloudToLaserScanNode::cloudCallback(
...
+    marker.pose.position.x = pcl_pointcloud.at(i).x;
+    marker.pose.position.y = pcl_pointcloud.at(i).y;
+    marker.pose.position.z = pcl_pointcloud.at(i).z - marker.scale.z * 0.5;
+    marker.pose.orientation.x = 0.0;
+    marker.pose.orientation.y = 0.0;
+    marker.pose.orientation.z = 0.0;
+    marker.pose.orientation.w = 1.0;
+    stixel_output.markers.push_back(marker);
+  }
+  stixel_viz_pub_->publish(stixel_output);
```
Comment (kenji-miyake) [inline src/pointcloud_to_laserscan_node.cpp]: IMO, it's better to move these marker-related lines to `unnamed namespace` or `member functions` or somewhere.

**Review by kenji-miyake — COMMENTED** [review]
I feel this is acceptable but contains some points to be improved.

**Comment by kenji-miyake** [comment]
@yukkysaito Are you okay to fix these points by @wep21 and me?

**Comment by yukkysaito** [comment]
@kenji-miyake off course :+1:

File: CMakeLists.txt (hunk 1 above)
Comment (wep21) [reply CMakeLists.txt]: Create PR on https://github.com/ros-perception/pointcloud_to_laserscan/pull/57

Comment (kenji-miyake) [reply CMakeLists.txt]: Super cool! :smile:

Comment (kenji-miyake) [reply CMakeLists.txt]: And could you please commit them to `tier4/main` and rebase this PR? :pray: 
Or will we wait for the response from upstream maintainers?

File: src/pointcloud_to_laserscan_node.cpp (hunk 4 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Const unique pointer is prohibited from being moved?
```
error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = sensor_msgs::msg::PointCloud2_<std::allocator<void> >; _Dp = std::default_delete<sensor_msgs::msg::PointCloud2_<std::allocator<void> > >]’
  270 |   pointcloud_pub_->publish(std::move(pointcloud_output_ptr));
```

Comment (kenji-miyake) [reply src/pointcloud_to_laserscan_node.cpp]: I'm sorry, okay let's leave it as it is!

File: src/pointcloud_to_laserscan_node.cpp (hunk 3 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Fixed at https://github.com/tier4/pointcloud_to_laserscan/pull/1/commits/f68d96d6d31cfc7fbc4531e14f53740dd8abad06

File: src/pointcloud_to_laserscan_node.cpp (hunk 2 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Fixed at https://github.com/tier4/pointcloud_to_laserscan/pull/1/commits/2035b1be800f727c607d91cc87d82345cd631398

File: src/pointcloud_to_laserscan_node.cpp (hunk 5 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Fixed at https://github.com/tier4/pointcloud_to_laserscan/pull/1/commits/eeed69727f86ca09ea631416fa4277562aad6154

File: src/pointcloud_to_laserscan_node.cpp (hunk 6 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Fixed at https://github.com/tier4/pointcloud_to_laserscan/pull/1/commits/3371ea3e610f4e4dfec523fdada45c11ada3d888

File: src/pointcloud_to_laserscan_node.cpp (hunk 7 above)
Comment (wep21) [reply src/pointcloud_to_laserscan_node.cpp]: Fixed at https://github.com/tier4/pointcloud_to_laserscan/pull/1/commits/e8e73626c8ff1c97512c4d71f5d0001b9fb4dc07

File: CMakeLists.txt (hunk 1 above)
Comment (wep21) [reply CMakeLists.txt]: I updated `tier4/main` and rebased this PR.

File: src/pointcloud_to_laserscan_node.cpp (hunk 6 above)
Comment (kenji-miyake) [reply src/pointcloud_to_laserscan_node.cpp]: Oh, `using` is nice!

**Review by kenji-miyake — DISMISSED** [review]
Clean code and diff! :smile: 
@wep21 Thank you so much!

File: launch/sample_pointcloud_to_laserscan_launch.py [hunk 8]
```diff
@@ -24,8 +24,16 @@ def generate_launch_description():
         ),
         Node(
             package='pointcloud_to_laserscan', executable='pointcloud_to_laserscan_node',
-            remappings=[('cloud_in', [LaunchConfiguration(variable_name='scanner'), '/cloud']),
-                        ('scan', [LaunchConfiguration(variable_name='scanner'), '/scan'])],
+            remappings=[('~/input/pointcloud',
+                        [LaunchConfiguration(variable_name='scanner'), '/cloud']),
```
Comment (kenji-miyake) [inline launch/sample_pointcloud_to_laserscan_launch.py]: Can remove `variable_name=`?

Comment (kenji-miyake) [reply launch/sample_pointcloud_to_laserscan_launch.py]: How about this?

```python
            remappings=[
                ('~/input/pointcloud', [LaunchConfiguration('scanner'), '/cloud']),
                ('~/output/laserscan', [LaunchConfiguration('scanner'), '/scan']),
                ('~/output/pointcloud', [LaunchConfiguration('scanner'), '/cloud_out']),
                ('~/output/ray', [LaunchConfiguration('scanner'), '/ray']),
                ('~/output/stixel', [LaunchConfiguration('scanner'), '/stixel'])
            ],
```

**Comment by kenji-miyake** [comment]
@wep21 Could you protect `tier4/main` or give me admin?

**Comment by wep21** [comment]
@kenji-miyake I gave you an admin privilege🙏

**Comment by wep21** [comment]
@yukkysaito @kenji-miyake Can I merge this PR?

**Comment by kenji-miyake** [comment]
@wep21 How about merge CI before merging this?

**Review by yukkysaito — DISMISSED** [review]

**Review by kenji-miyake — APPROVED** [review]

**Pull request merged**